### DISCUSSION TITLE
Build topology with language like python?

### DISCUSSION THREAD

**Discussion by lutong-z** [sparcians/map#413] [Ideas]
Hi, I'm trying to reconstruct my RISC-V performance model with sparta due to its well-developed visualization. 

A big concern is that there is not a programable language to build topology at runtime. I notice that the CoreExample in this project use yaml to configure parameter and directly build the topology in c++ and I used yaml to build CPU's topology in my former proejct which is  a total mess with more complex design.

The reason of being able to build topology in run time is that we can make the interconnecting of NoC more flexible and enable tuning of CPU(like nubmer of the ALU,LD).

**Comment by klingaard** [comment]
I completely understand where you're coming from.  A couple of comments regarding this:

### Python Support
 Sparta's original design and development was driven by the need to build topologies at runtime, specifically using python.  This is why you cannot name units, resources, parameters with special characters in them (conflicts with python naming rules).  There is a python-enabled build of sparta, and it does work, but is very incomplete.  There are some rudimentary build directions [here](https://github.com/sparcians/map/tree/master/sparta#python-shell).   The _eventual_ intent is to do something like this (in python):
```
./my_sim --python-shell setup_mysim.py
```
where the `setup_mysim.py` could look something like this:
```
sim = MySim()
fetch = sim.createFetch()
fetch.params.num_to_fetch = 3
decode = sim.createDecode()
decode.params.num_to_decode = 3

sim.enterConfiguration()
sparta.bind(fetch.ports.out_insts, decode.ports.in_insts)
sim.enterFinalized()
sim.run()
sim.enterTeardown()
```
### Using a YAML file
I know you mentioned how this was an ugly mess in your previous company, but it might be a _little_ better here?  You be the judge.  There's an example of how to use the Sparta extensions to define and initialize topologies in the RISC-V Performance Model repository.  Documentation is [here](https://github.com/riscv-software-src/riscv-perf-model#experimenting-with-architectures).

By using extensions, you can define things like number of ALUs, number of FPUs, number of LSUs, number of issue queues, types of functional units, etc.  Your imagination is the limit.  There's an example of how to "slurp" in the topology extension in [Execute.cpp](https://github.com/riscv-software-src/riscv-perf-model/blob/master/core/Execute.cpp) to setup downstream execution units.  Binding these units together is also generic and is found [here](https://github.com/riscv-software-src/riscv-perf-model/blob/77afb4c6fc39bdc844bf1cd481a0e5661f886462/core/CPUTopology.cpp#L244).   

I hope this will get your mental wheels spinning on how to make topologies work while still sticking with the same code base.  Of course, if you're interested in working on the python implementation, I'd love the help!

**Reply by lutong-z** [reply]
Got it. I have checked olympia's implementation and it looks good, but I still tend to build topologies with python. The reasons are : 

**For design modeling and configuration checking**
The tool we want to develop is a cycle accurate model which can directly guide RTL implementation, so we will have a more complex topology like "issue arbitration" "port sharing" "complex bypass networks" and so on. We can also have a clean configuration with archtecture specializd in c++ like implementing "instruction scheduling layer", "execution layer". For me, I want to have more component and do the connection task at runtime. (Think about integrating some RTL modules with model  :D)
So there should be a checking tool to help human with debuging. For example, after we create a configuration and for human's reason, there are some disconnections of function units which result in that the machine can't execute all the operation which it should support(ISA). 
If we build topologies with python, it's possible to check the error with static information and give some hints to help users create new system.

**YAML is not scalable enough**
YAML is easy to understand, but it's not scalable. Think about a system with hundred of core and we are doing architecture exploration based on a mesh design. We duplicate the processor and connect them togother. We can implement the duplication task in c++ or YAML or python. The c++ one needs c++ code modification. The YAML one has too many informations which is hard to describe and check for human. I think it's better to put something together and get a bigger one. Then connect.

**Visualization**
If we want to generate the archtecture diagram with the topology, it's easier when we maintain topology descriptions in python. Those information can also be the visualization meta of helio.
Another building flow can be building the components block diagram and connect them with GUI to generate topology and the simulator.


Thanks for your excelent work and we are glad to make some contribution on python implementation.

**Comment by ghost** [comment]
Wow, your use cases would take the Sparta modeling framework to the next level!  

Totally agree that python is the way to go (and that YAML is NOT scalable for this); I only wish I had the time/focus to work on it.  For massive simulation projects, we did enable support for glob pattern matching to support the setting of multiple components:
```
top.core_complex.core*.alu?.params.num_pips: 2
```
But this barely mitigates the issue.

To really coalesce the python effort for Sparta, I'd start almost from scratch on the current design.  Cython or Boost Python are acceptable bridges, but if there are better packages for this, I'm open to ideas.  Also, I'd like to hash out the full requirements for this extension and build a plan for development.  I figured I should document what is already there and state what the intentions where.  Thoughts?

Finally, you're absolutely correct in the idea that with a python interface, generation of architectural diagrams and layouts (for tools like Argos, etc) are simpler to create.  And a GUI drag-and-drop design is definitely within reach.
